### ISSUE TITLE
Is there a backup functionality of Starcounter?

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#3356]
The question came from one of our users, @jwillmer. The context is latest nightly **pnext** as of today.

**Comment by jwillmer** [comment]
Background: We like to test some scenarios and we would like to create a default database with some records and use this database for our tests. That means that we would like to revert after each test to the prepared database.

**Comment by dan31** [comment]
@jwillmer you are now using a version "pnext". Inside this version there is an ongoing work on what we call "log retention policy", which includes typical backup scenarios and also caters for replication needs. Once available, we will supply you with documentation on the feature. Until that, the generic way to backup database images is to use a simple tooling atop Windows Shadow Copy as described in http://starcounter.io/guides/tools/run-starcounter-in-production/. You might be already familiar with the described techniques, eve though I recommend checking the article.

For your scenario, one option is to use Unload/Reload facility, as described in http://starcounter.io/guides/tools/staradmin/. This permits to migrate data between incompatible versions of Starcounter as well. More straightforward way is to copy a database image and substitute the current image file with it.

**Comment by k-rus** [comment]
> We like to test some scenarios and we would like to create a default database with some records and use this database for our tests. That means that we would like to revert after each test to the prepared database.

In this case the simplest way is to copy the folder of the prepared database or server somewhere, and then after an experiment revert the database by replacing it with the copy. I.e., just remove the folder and copy the prepared database folder into the place. I believe it should work.

**Comment by jwillmer** [comment]
@k-rus  That seams simple enough. Can I use this strategy for production backups as well or is there a problem if the database is running (invalid states)?

**Comment by k-rus** [comment]
> @k-rus That seams simple enough. Can I use this strategy for production backups as well or is there a problem if the database is running (invalid states)?

I don't think you can copy files while database is running due to locked files, but I can be wrong. In any case it will slow down your disc considerably, which might affect availability of your production system.
For production backup it is better to use @dan31 suggestion:


> the generic way to backup database images is to use a simple tooling atop Windows Shadow Copy as described in http://starcounter.io/guides/tools/run-starcounter-in-production/.

**Comment by dan31** [comment]
In 2.1 and further on you will not be let copy active image files directly, because they might be in the update state thus locked, as @k-rus also points. Some will also be locked when the database is not writing to the files. Even if they aren't locked, in a progress of copying such files you will get a copy of a database image in presumably inconsistent state, since your copying spans time interval but not a point-in-time state of the files, hence you might get copies of files updated at different times. The proposed strategy with a Shadow Copy script (section "Hot backup of database files using VSS" there in http://starcounter.io/guides/tools/run-starcounter-in-production, the solution is explained there in detail) will give you a point-in-time slice of the image files, which you can safely copy to the other place. VSS service will not bring to any noticeable slowdown, it has no "stop-the-world" drawback, it works on a block-level and keeps the snapshot needed for a backup copy only short period in time.

In addition to that, the intended log retention tool will provide the same consistent backing up, also taking into account log files retention in respect to replication. I will not go deeper into that in a current issue, I would prefer to explain the topic deeper when we run the replication scenario together and, even better, have some offline session together and have a tool in-place.
